src/BreakpointViewer.cpp

```diff
@@ -0,0 +1,621 @@
+	} else {
+		type2 = Breakpoints::BREAKPOINT;
+	}
+
+	// qDebug() << "createBreakpoint, type" << type << "row" << row;
+	// for (auto item: items) {
+	// 	qDebug() << "item" << item.type << item.row;
+	// }
+
+	bool ok;
+	int address, endRange;
+	QString location = table->item(row, LOCATION)->text();
+	ok = processLocationField(-1, type, location, address, endRange);
+	if (!ok) {
+		auto enabledItem = table->item(row, ENABLED);
+		userMode = false;
+		enabledItem->setCheckState(Qt::Unchecked);
+		userMode = true;
+		return;
+	}
+
+	QString condition = table->item(row, CONDITION)->text();
+
+	qint8 ps, ss;
+	ok = processSlotField(-1, table->item(row, SLOT)->text(), ps, ss);
+	assert(ok);
+
+	qint16 segment;
+	ok = processSegmentField(-1, table->item(row, SEGMENT)->text(), segment);
+	assert(ok);
+
+	const QString cmdStr = Breakpoints::createSetCommand(type2, address, ps, ss, segment,
+			endRange, condition);
+	auto command = new Command(cmdStr,
+		[this] (const QString result) {
+			emit contentsUpdated(false);
+		}
+	);
+	CommClient::instance().sendCommand(command);
+}
+
+// TODO: move the createRemoveCommand to a session manager
+void BreakpointViewer::replaceBreakpoint(Type type, int row)
+{
+	// qDebug() << "replaceBreakpoint() called";
+	int index = items.indexOf({type, row});
+	// qDebug() << "replaceBreakpoint, type" << type << "row" << row;
+	// for (auto item: items) {
+	// 	qDebug() << "item" << item.type << item.row;
+	// }
+	assert(index != -1);
+
+	const QString id = breakpoints->getAttribute(index, "id").toString();
+	const QString cmdStr = breakpoints->createRemoveCommand(id);
+
+	qDebug() << "cmdStr =" << cmdStr;
+	auto command = new Command(cmdStr,
+		[this, type, row] (const QString& result) {
+			qDebug() << "result =" << result;
+			createBreakpoint(type, row);
+		}
+	);
+	CommClient::instance().sendCommand(command);
+}
+
+// TODO: move the createRemoveCommand to a session manager
+void BreakpointViewer::removeBreakpoint(Type type, int row)
+{
+	int index = items.indexOf({type, row});
+	// qDebug() << "removeBreakpoint, type" << type << "row" << row;
+	// for (auto item: items) {
+	// 	qDebug() << "item" << item.type << item.row;
+	// }
+	// assert(index != -1);
+
+	QString id = breakpoints->getAttribute(index, "id").toString();
+	QString cmdStr = Breakpoints::createRemoveCommand(id);
+
+	auto command = new Command(cmdStr,
+		[this] (const QString result) {
+			emit contentsUpdated(false);
+	});
+	CommClient::instance().sendCommand(command);
+}
+
+bool BreakpointViewer::processSlotField(int index, const QString& field, qint8& ps, qint8& ss)
+{
+	QStringList s = field.simplified().split("/", Qt::SplitBehaviorFlags::SkipEmptyParts);
+
+	if (s[0].compare("X", Qt::CaseInsensitive)) {
+		bool ok;
+		int value = s[0].toInt(&ok);
+		if (!ok || (value < 0 && value > 3)) {
+			// restore value
+			if (index != -1)
+				ps = breakpoints->getAttribute(index, "ps").toInt();
+			else
+				return false;
```

Comment:
(detail, style)
Prefer braces in multi-line if/else-if/else statements, even if there's only a single statement in some branch.
I'll only mention this once, but I saw it in some other places as well.